### ISSUE TITLE
[BUG perl 5.10.0] Permission denied when building as non-root user on FreeBSD 6.2+

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9454]
Migrated from [rt.perl.org#58180](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=58180) (status was 'resolved')

Searchable as RT58180$

**Comment by p5pRT** [comment]
### From italpro@qs1213.pair.com

#### Created by italpro@qs1213.pair.com



When building Perl 5\.10\.0 under FreeBSD 6\.2 \(and I think more recent
versions\) as well as a non\-root \(unprivilged user\)\.

This happens because recent FreeBSD releases make /usr/home unreadable
by regular users\, and for some reasons perl installation fails if it
cannon opendir\(\) that directory \- and it probably shouldn't fail\.
I asked the superuser to make that directory readable to me and perl
built and installed regularly\.

==> HOW TO REPRODUCE

\- Login as unprivileged user
\- Unpack the perl 5\.10 distribution
\# sh Configure \-de \-Dprefix=/your/local/directory
\# make

==> ERROR \(reporting the last part of "make" output\)

cd lib/unicore &&   \.\./\.\./miniperl \-I\.\./\.\./lib mktables \-w
touch uni\.data
\`sh  cflags "optimize='\-O'" perlmain\.o\`  perlmain\.c
&nbsp;      CCCMD =  cc \-DPERL\_CORE \-c \-DHAS\_FPSETMASK \-DHAS\_FLOATINGPOINT\_H \-fno\-strict\-aliasing \-pipe \-I/usr/local/include  \-std=c89 \-O
\+\-Wall \-ansi   \-W \-Wextra \-Wdeclaration\-after\-statement \-Wendif\-labels

&nbsp;    Making DynaLoader \(static\)
Makefile out\-of\-date with respect to \.\./\.\./lib/Config\.pm \.\./\.\./config\.h
Cleaning current config before rebuilding Makefile\.\.\.
make \-f Makefile\.old clean > /dev/null 2>&1
\.\./\.\./miniperl "\-I\.\./\.\./lib" "\-I\.\./\.\./lib" Makefile\.PL "INSTALLDIRS=perl" "INSTALLMAN3DIR=none" "PERL\_CORE=1" "LIBPERL\_A=libperl\.a"
opendir\(\./\.\./\.\./\.\./\.\./\.\.\)&#8203;: Permission denied at \.\./\.\./lib/File/Spec/Unix\.pm line 478
opendir\(\./\.\./\.\./\.\./\.\./\.\.\)&#8203;: Permission denied at \.\./\.\./lib/File/Spec/Unix\.pm line 478
Writing Makefile for DynaLoader
==> Your Makefile has been rebuilt\. \<==
==> Please rerun the make command\.  \<==
false
\*\*\* Error code 1

Stop in /usr/home/italpro/soft/perl\-5\.10\.0/ext/DynaLoader\.
make config failed\, continuing anyway\.\.\.
/miniperl&#8203;: not found
\*\*\* Error code 127

Stop in /usr/home/italpro/soft/perl\-5\.10\.0/ext/DynaLoader\.
\*\*\* Error code 1

<details><summary>Perl Info</summary>

```

Flags:
    category=install
    severity=medium

Site configuration information for perl 5.10.0:

Configured by italpro at Wed Aug 20 16:57:09 EDT 2008.

Summary of my perl5 (revision 5 version 10 subversion 0) configuration:
  Platform:
    osname=freebsd, osvers=6.2-release-p1, archname=i386-freebsd
    uname='freebsd qs1213.pair.com 6.2-release-p1 freebsd 6.2-release-p1 #1: mon may 7 14:30:09 edt 2007 mlehner@mayon.pair.com:usrobjusrsrcsys6pairc-smp i386 '
    config_args='-de -Dprefix=/usr/home/italpro/localperl -Dinstallscript=/usr/home/italpro/localperl/scripts'
    hint=previous, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -fno-strict-aliasing -pipe -I/usr/local/include',
    optimize='-O',
    cppflags='-DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -fno-strict-aliasing -pipe -I/usr/local/include -DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='3.4.6 [FreeBSD] 20060305', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags ='-Wl,-E  -L/usr/local/lib'
    libpth=/usr/lib /usr/local/lib
    libs=-lgdbm -lm -lcrypt -lutil -lc
    perllibs=-lm -lcrypt -lutil -lc
    libc=, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-DPIC -fPIC', lddlflags='-shared  -L/usr/local/lib'

Locally applied patches:
    


@INC for perl 5.10.0:
    /usr/home/italpro/localperl/lib/5.10.0/i386-freebsd
    /usr/home/italpro/localperl/lib/5.10.0
    /usr/home/italpro/localperl/lib/site_perl/5.10.0/i386-freebsd
    /usr/home/italpro/localperl/lib/site_perl/5.10.0
    .


Environment for perl 5.10.0:
    HOME=/usr/home/italpro
    LANG (unset)
    LANGUAGE (unset)
    LC_ALL=en_US.UTF-8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/italpro/bin:/home/italpro/localperl/bin:/home/italpro/localperl/scripts:/home/italpro/local/bin:/sbin:/bin:/usr/sbin:/usr/bin:/usr/games:/usr/local/sbin:/usr/local/bin:/usr/X11R6/bin:/usr/home/italpro/bin
    PERL_BADLANG (unset)
    SHELL=/usr/local/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From tsibley@cpan.org

I can confirm that this happens for me with both 5\.10\.0 and 5\.8\.8 on
Linux \(2\.4\.32\) where /home is unreadable \(but executable\) by my user\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @smpeters

On Wed Aug 20 14&#8203;:15&#8203;:07 2008\, italpro@&#8203;qs1213\.pair\.com wrote&#8203;:
> 
> This is a bug report for perl from italpro@&#8203;qs1213\.pair\.com\,
> generated with the help of perlbug 1\.36 running under perl 5\.10\.0\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> When building Perl 5\.10\.0 under FreeBSD 6\.2 \(and I think more recent
> versions\) as well as a non\-root \(unprivilged user\)\.
> 
> This happens because recent FreeBSD releases make /usr/home unreadable
> by regular users\, and for some reasons perl installation fails if it
> cannon opendir\(\) that directory \- and it probably shouldn't fail\.
> I asked the superuser to make that directory readable to me and perl
> built and installed regularly\.
> 
> ==> HOW TO REPRODUCE
> 
> \- Login as unprivileged user
> \- Unpack the perl 5\.10 distribution
> \# sh Configure \-de \-Dprefix=/your/local/directory
> \# make
> 
> ==> ERROR \(reporting the last part of "make" output\)
> 
> cd lib/unicore &&   \.\./\.\./miniperl \-I\.\./\.\./lib mktables \-w
> touch uni\.data
> \`sh  cflags "optimize='\-O'" perlmain\.o\`  perlmain\.c
>       CCCMD =  cc \-DPERL\_CORE \-c \-DHAS\_FPSETMASK \-DHAS\_FLOATINGPOINT\_H
>    \-fno\-strict\-aliasing \-pipe \-I/usr/local/include  \-std=c89 \-O
> \+\-Wall \-ansi   \-W \-Wextra \-Wdeclaration\-after\-statement \-Wendif\-labels
> 
>     Making DynaLoader \(static\)
> Makefile out\-of\-date with respect to \.\./\.\./lib/Config\.pm
>    \.\./\.\./config\.h
> Cleaning current config before rebuilding Makefile\.\.\.
> make \-f Makefile\.old clean > /dev/null 2>&1
> \.\./\.\./miniperl "\-I\.\./\.\./lib" "\-I\.\./\.\./lib" Makefile\.PL
>    "INSTALLDIRS=perl" "INSTALLMAN3DIR=none" "PERL\_CORE=1"
>    "LIBPERL\_A=libperl\.a"
> opendir\(\./\.\./\.\./\.\./\.\./\.\.\)&#8203;: Permission denied at
>    \.\./\.\./lib/File/Spec/Unix\.pm line 478
> opendir\(\./\.\./\.\./\.\./\.\./\.\.\)&#8203;: Permission denied at
>    \.\./\.\./lib/File/Spec/Unix\.pm line 478
> Writing Makefile for DynaLoader
> ==> Your Makefile has been rebuilt\. \<==
> ==> Please rerun the make command\.  \<==
> false
> \*\*\* Error code 1
> 
> Stop in /usr/home/italpro/soft/perl\-5\.10\.0/ext/DynaLoader\.
> make config failed\, continuing anyway\.\.\.
> /miniperl&#8203;: not found
> \*\*\* Error code 127
> 
> Stop in /usr/home/italpro/soft/perl\-5\.10\.0/ext/DynaLoader\.
> \*\*\* Error code 1
> 
> 

What are your exact permissions on /usr/home?  I want to investigate this problem\, but I'd 
like to work with the exact permissions you have\.

Thanks\,

Steve Peters

**Comment by p5pRT** [comment]
### From arthas@nordest.pm.org

Hello Steve\!

This was on the FreeBSD box of my ISP\. I am a non\-root user and I
couldn't even
read /usr/home \(I could only read my home directory at
/usr/home/italpro\)\. This
seems to be common on FreeBSD 6\.x\, where normal user can't see /usr/home
contents in order not to see the list of the other user directories\.

I then asked and obtained read access to that directory from my ISP\, and
perl
built fine\. I however feel that perl should build locally even if
/usr/home is
not accessible\.

Thanks in advance for investigating\.

Michele\.

**Comment by p5pRT** [comment]
### From @lordarthas

Hi Steve\!

> What are your exact permissions on /usr/home?  I want to investigate this problem\, but I'd 
> like to work with the exact permissions you have\.

This was on the FreeBSD box of my ISP\. I am a non\-root user and I couldn't even
read /usr/home \(I could only read my home directory at /usr/home/italpro\)\. This
seems to be common on FreeBSD 6\.x\, where normal user can't see /usr/home
contents in order not to see the list of the other user directories\.

I then asked and obtained read access to that directory from my ISP\, and perl
built fine\. I however feel that perl should build locally even if /usr/home is
not accessible\.

Thanks in advance for investigating\.

Talk to you soon\,
Michele\.

\-\- 
Michele Beltrame
http&#8203;://www\.italpro\.net/
ICQ 76660101 \- SkyPe arthas77

**Comment by p5pRT** [comment]
### From schmonz@schmonz.com

Like tsibley\, I can confirm this with 5\.10\.0 on Linux 2\.4\.32 \(as
provided by DreamHost\) with /home thus&#8203;:

drwxrws\-\-x  394 root staff 36864 2009\-04\-24 09&#8203;:17 /home

Any progress on a fix?

Thanks\,

\- Amitai

**Comment by p5pRT** [comment]
### From i@h2l.name

I have the same problem when tried to install 5\.10\.0
Is there any progress about the issue?

Some information about the system I am using&#8203;:

$ cat /etc/issue
Debian GNU/Linux 3\.1 \\n \\l

$ uname \-a
Linux razzle
2\.4\.32\-grsec\+f6b\+gr217\+nfs\+a32\+fuse23\+tg\+\+\+opt\+c8\+gr2b\-v6\.194 \#1 SMP Tue
Jun 6 15&#8203;:52&#8203;:09 PDT 2006 i686 GNU/Linux

$ ls \-dl /home
drwxrws\-\-x  234 root staff 28672 2009\-05\-20 12&#8203;:47 /home/

**Comment by p5pRT** [comment]
### From p5p@spam.wizbit.be

On Wed Aug 20 14&#8203;:15&#8203;:07 2008\, italpro@&#8203;qs1213\.pair\.com wrote&#8203;:
> 
> This is a bug report for perl from italpro@&#8203;qs1213\.pair\.com\,
> generated with the help of perlbug 1\.36 running under perl 5\.10\.0\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> When building Perl 5\.10\.0 under FreeBSD 6\.2 \(and I think more recent
> versions\) as well as a non\-root \(unprivilged user\)\.
> 
> This happens because recent FreeBSD releases make /usr/home 
unreadable
> by regular users\, and for some reasons perl installation fails if it
> cannon opendir\(\) that directory \- and it probably shouldn't fail\.
> I asked the superuser to make that directory readable to me and perl
> built and installed regularly\.
> 
> ==> HOW TO REPRODUCE
> 
> \- Login as unprivileged user
> \- Unpack the perl 5\.10 distribution
> \# sh Configure \-de \-Dprefix=/your/local/directory
> \# make
> 
> ==> ERROR \(reporting the last part of "make" output\)
> 
> cd lib/unicore &&   \.\./\.\./miniperl \-I\.\./\.\./lib mktables \-w
> touch uni\.data
> \`sh  cflags "optimize='\-O'" perlmain\.o\`  perlmain\.c
>       CCCMD =  cc \-DPERL\_CORE \-c \-DHAS\_FPSETMASK \-
DHAS\_FLOATINGPOINT\_H
>    \-fno\-strict\-aliasing \-pipe \-I/usr/local/include  \-std=c89 \-O
> \+\-Wall \-ansi   \-W \-Wextra \-Wdeclaration\-after\-statement \-Wendif\-
labels
> 
>     Making DynaLoader \(static\)
> Makefile out\-of\-date with respect to \.\./\.\./lib/Config\.pm
>    \.\./\.\./config\.h
> Cleaning current config before rebuilding Makefile\.\.\.
> make \-f Makefile\.old clean > /dev/null 2>&1
> \.\./\.\./miniperl "\-I\.\./\.\./lib" "\-I\.\./\.\./lib" Makefile\.PL
>    "INSTALLDIRS=perl" "INSTALLMAN3DIR=none" "PERL\_CORE=1"
>    "LIBPERL\_A=libperl\.a"
> opendir\(\./\.\./\.\./\.\./\.\./\.\.\)&#8203;: Permission denied at
>    \.\./\.\./lib/File/Spec/Unix\.pm line 478
> opendir\(\./\.\./\.\./\.\./\.\./\.\.\)&#8203;: Permission denied at
>    \.\./\.\./lib/File/Spec/Unix\.pm line 478
> Writing Makefile for DynaLoader
> ==> Your Makefile has been rebuilt\. \<==
> ==> Please rerun the make command\.  \<==
> false
> \*\*\* Error code 1
> 
> Stop in /usr/home/italpro/soft/perl\-5\.10\.0/ext/DynaLoader\.
> make config failed\, continuing anyway\.\.\.
> /miniperl&#8203;: not found
> \*\*\* Error code 127
> 
> Stop in /usr/home/italpro/soft/perl\-5\.10\.0/ext/DynaLoader\.
> \*\*\* Error code 1
> 
> 
> \[Please do not change anything below this line\]
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

This problem was fixed in perl\-5\.10\.1 \(and perl\-5\.11\.x\)\.

The cause&#8203;:
http&#8203;://rt\.cpan\.org/Public/Bug/Display\.html?id=34971
Bug \#34971 for ExtUtils\-MakeMaker&#8203;: ExtUtils&#8203;::MakeMaker Requires All 
Parent

The fix&#8203;:
http&#8203;://cpansearch\.perl\.org/src/SMUELLER/PathTools\-3\.31/Changes

3\.26 \- Sun Jan 13 21&#8203;:59&#8203;:20 2008

 \- case\_tolerant\(\) on Cygwin will now avoid a painful death when
&nbsp;   Cygwin&#8203;::mount\_flags\(\) isn't defined\, as is the case for perl \<
&nbsp;   5\.10\.  It will now just return 1\, which is what it always did
&nbsp;   before it got so smart\. \[Spotted by Emanuele Zeppieri\]

 \- abs\_path\(\) on Unix\(ish\) platforms has been upgraded to a much later
&nbsp;   version of the underlying C code from BSD\. \[Michael Schwern\]

The diff&#8203;:
@&#8203;@&#8203; \-540\,8 \+540\,8 @&#8203;@&#8203;
&nbsp;        local \*PARENT;
&nbsp;        unless \(opendir\(PARENT\, $dotdots\)\)
&nbsp;        \{
\-           \_carp\("opendir\($dotdots\)&#8203;: $\!"\);
\-           return '';
\+           \# probably a permissions issue\.  Try the native command\.
\+           return File&#8203;::Spec\->rel2abs\( $start\, \_backtick\_pwd\(\) \);
&nbsp;        \}



Best regards\,

Bram

**Comment by p5pRT** [comment]
p5p@spam.wizbit.be - Status changed from 'open' to 'resolved'
